### PR TITLE
[WIP] Workaround and fix for vSphere CC

### PR DESCRIPTION
**Pull request by 12345lcr**
### Which issue(s) this PR fixes

Fixes #

### Release note

```release-note

```

### CODE REVIEW

File: addons/controllers/cpi/vspherecpiconfig_utils.go [hunk 1]
```diff
@@ -36,15 +35,15 @@ func (r *VSphereCPIConfigReconciler) mapCPIConfigToDataValuesNonParavirtual( //
...
+	// get the vcenter values
+	vcenterValues, err := util.ParseVCenterValues(cluster)
 	if err != nil {
 		return nil, err
 	}
 
 	// derive the thumbprint, server from the vsphere cluster object
-	d.TLSThumbprint = vsphereCluster.Spec.Thumbprint
-	d.Server = vsphereCluster.Spec.Server
+	d.TLSThumbprint = vcenterValues.TlsThumbprint
```
Comment (vijaykatam) [inline addons/controllers/cpi/vspherecpiconfig_utils.go:45]: Question: Why not continue to rely on VsphereCluster object?

File: addons/controllers/cpi/vspherecpiconfig_utils.go [hunk 2]
```diff
@@ -84,27 +71,9 @@ func (r *VSphereCPIConfigReconciler) mapCPIConfigToDataValuesNonParavirtual( //
 		d.NoProxy = cluster.Annotations[pkgtypes.NoProxyConfigAnnotation]
 	}
 
-	// derive nsxt related configs from cluster variable
-	d.Nsxt.PodRoutingEnabled = r.tryParseClusterVariableBool(cluster, NsxtPodRoutingEnabledVarName)
```
Comment (vijaykatam) [inline addons/controllers/cpi/vspherecpiconfig_utils.go:88]: Why are nsxt values being removed?

File: addons/controllers/csi/vspherecsiconfig_utils.go [hunk 3]
```diff
@@ -122,11 +113,6 @@ func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesNonParavirtu
 		return nil, err
 	}
 
-	dvs.VSphereCSI.Region = cutil.TryParseClusterVariableString(ctx, cluster, VSphereRegionVarName)
```
Comment (vijaykatam) [inline addons/controllers/csi/vspherecsiconfig_utils.go:125]: These shouldn't be removed.

File: addons/controllers/cpi/vspherecpiconfig_utils.go (hunk 1 above)
Comment (DanielXiao) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:45]: It is just easier to retrieve all information from the variable vcenter. Otherwise we need to query VsphereCluster and VsphereMachineTemplate. If this is risky, we can go back to VsphereCluster.

File: addons/controllers/cpi/vspherecpiconfig_utils.go (hunk 2 above)
Comment (DanielXiao) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:88]: These values are not required by cluster-api, so they will never be cluster variables.

File: addons/controllers/csi/vspherecsiconfig_utils.go (hunk 3 above)
Comment (DanielXiao) [reply addons/controllers/csi/vspherecsiconfig_utils.go:125]: The same as above.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml [hunk 4]
```diff
@@ -541,7 +534,7 @@ spec:
       - op: replace
         path: /spec/template/spec/template
         valueFrom:
-          template: '{{ (index .TKR_DATA .builtin.controlPlane.version).osImageRef.template }}'
+          template: '/dc0/vm/photon-3-kube- {{- (index .TKR_DATA .builtin.controlPlane.version).osImageRef.version }}'
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml:537]: We will want to revert this change before committing. This is necessary until this #2748 is merged.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml [hunk 5]
```diff
@@ -605,7 +598,7 @@ spec:
       - op: replace
         path: /spec/template/spec/template
         valueFrom:
-          template: '{{ (index .TKR_DATA .builtin.machineDeployment.version).osImageRef.template }}'
+          template: '/dc0/vm/photon-3-kube- {{- (index .TKR_DATA .builtin.machineDeployment.version).osImageRef.version }}'
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml:601]: this will also need to be reverted.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml [hunk 6]
```diff
@@ -879,26 +872,8 @@ spec:
...
-          template: |
-            {{ $first := true }}
-            {{- range $key, $val := (index .TKR_DATA .builtin.controlPlane.version).labels }}
-            {{- if $first }}
-              {{- $first = false }}
-            {{- else -}}
-              ,
-            {{- end }}
-            {{- $key -}} = {{- $val }}
-            {{- end}}
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml:901]: This will need to be reverted and enhanced. The $val variable can only contain alphanumeric characters, `-`, `_`, and `.`. We will need to filter out any label values that contain disallowed characters.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml [hunk 7]
```diff
@@ -879,26 +872,8 @@ spec:
         path: /spec/template/spec/kubeadmConfigSpec/postKubeadmCommands/-
         value: sed -i '/listen-client-urls/ s/$/,https:\/\/127.0.0.1:2379/' /etc/kubernetes/manifests/etcd.yaml
   - name: nodeLabels
+    enabledIf: '{{ empty .nodePoolLabels | not }}'
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml:875]: this can be reverted because the change below should also be reverted.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml [hunk 8]
```diff
@@ -912,15 +887,6 @@ spec:
...
           template: |
-            {{ $first := true }}
-            {{- range $key, $val := (index .TKR_DATA .builtin.machineDeployment.version).labels }}
-            {{- if $first }}
-              {{- $first = false }}
-            {{- else -}}
-              ,
-            {{- end }}
-            {{- $key -}} = {{- $val }}
-            {{- end}}
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml:923]: this needs to be reverted and enhanced as in the above

File: addons/controllers/cpi/vspherecpiconfig_utils.go (hunk 2 above)
Comment (vijaykatam) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:88]: The file you referenced is for backwards compatibility with legacy configuration. Going forward we expect these to be provided as cluster variables, if they are missing then we have a gap in the cluster class definition. cc @vuil @lubronzhan

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay.yaml [hunk 9]
```diff
@@ -14,17 +14,6 @@ spec:
...
-      #@ if data.values.ENABLE_MHC or data.values.ENABLE_MHC_CONTROL_PLANE:
-      nodeStartupTimeout: #@ data.values.NODE_STARTUP_TIMEOUT
-      unhealthyConditions:
-      - type: Ready
-        status: Unknown
-        timeout: #@ data.values.MHC_UNKNOWN_STATUS_TIMEOUT
-      - type: Ready
-        status: "False"
-        timeout: #@ data.values.MHC_FALSE_STATUS_TIMEOUT
-      #@ end
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay.yaml:27]: This should be reverted, but also add an overlay match to add the whole machineHealthCheck. `machineHealthCheck` can't be left empty. So the whole block must be added, or none at all.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay.yaml [hunk 10]
```diff
@@ -33,17 +22,6 @@ spec:
...
-        #@ if data.values.ENABLE_MHC or data.values.ENABLE_MHC_WORKER_NODE:
-        nodeStartupTimeout: #@ data.values.NODE_STARTUP_TIMEOUT
-        unhealthyConditions:
-        - type: Ready
-          status: Unknown
-          timeout: #@ data.values.MHC_UNKNOWN_STATUS_TIMEOUT
-        - type: Ready
-          status: "False"
-          timeout: #@ data.values.MHC_FALSE_STATUS_TIMEOUT
-        #@ end
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay.yaml:46]: Same as above.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay.yaml [hunk 11]
```diff
@@ -56,17 +34,6 @@ spec:
...
```
Comment (tenczar) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay.yaml:69]: same as above

File: pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml [hunk 12]
```diff
@@ -43,9 +40,15 @@ spec:
...
-    version: #@ data.values.KUBERNETES_VERSION
+    #! version: #@ data.values.KUBERNETES_VERSION
+    #! hack to match up tkr
+    version: #@ "{}-tkg.1-zshippable".format(data.values.KUBERNETES_VERSION)
     controlPlane:
       replicas: #@ data.values.CONTROL_PLANE_MACHINE_COUNT
+      #@overlay/match missing_ok=True
+      metadata:
+        annotations:
+          run.tanzu.vmware.com/resolve-os-image: image-type=ova,os-name=photon
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml:51]: we will need to populate the `resolve-os-image` dynamically. that can be saved for a separate issue for later.

File: pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml [hunk 13]
```diff
@@ -54,6 +57,10 @@ spec:
         name: md-0
         #@overlay/match missing_ok=True
         replicas: #@ data.values.WORKER_MACHINE_COUNT
+        #@overlay/match missing_ok=True
+        metadata:
+          annotations:
+            run.tanzu.vmware.com/resolve-os-image: image-type=ova,os-name=photon
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml:63]: same as above

File: pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml [hunk 14]
```diff
@@ -63,13 +70,21 @@ spec:
       - class: tkg-worker
         name: md-1
         replicas: #@ data.values.WORKER_MACHINE_COUNT_1
+        #@overlay/match missing_ok=True
+        metadata:
+          annotations:
+            run.tanzu.vmware.com/resolve-os-image: image-type=ova,os-name=photon
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml:76]: same as above

File: pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml [hunk 15]
```diff
@@ -63,13 +70,21 @@ spec:
...
         failureDomain: #@ data.values.VSPHERE_AZ_1
         #@ end
       #@overlay/append
       - class: tkg-worker
         name: md-2
         replicas: #@ data.values.WORKER_MACHINE_COUNT_2
+        #@overlay/match missing_ok=True
+        metadata:
+          annotations:
+            run.tanzu.vmware.com/resolve-os-image: image-type=ova,os-name=photon
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml:87]: same as above

File: pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml [hunk 16]
```diff
@@ -80,6 +95,11 @@ spec:
       - class: tkg-worker-windows
         name: md-0
         replicas: #@ data.values.WORKER_MACHINE_COUNT
+        #@overlay/match missing_ok=True
+        metadata:
+          annotations:
+            #! VVV TODO(tenczar) os-name handling
+            run.tanzu.vmware.com/resolve-os-image: image-type=ova,os-name=windows
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-vsphere/v1.3.1/yttcc/overlay.yaml:102]: same but more complicated because this is windows

File: pkg/v1/providers/yttcb/clusterbootstrap.yaml [hunk 17]
```diff
@@ -181,6 +181,7 @@ metadata:
   annotations:
     tkg.tanzu.vmware.com/add-missing-fields-from-tkr: #@ data.values.KUBERNETES_RELEASE
 spec:
+  #@ if data.values.CNI != "antrea" or antrea_configs_exist():
```
Comment (tenczar) [inline pkg/v1/providers/yttcb/clusterbootstrap.yaml:184]: We'll need to revisit this. I don't think the antrea_configs_exist() method is accurate. this can be a separate PR. we'll need to file and issue for it.

**Review by tenczar — CHANGES_REQUESTED** [review]
some comments that need to be addressed. I'm leaving the review of the addon code to Vijay and Leo, make sure you consult with them before merging.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml (hunk 4 above)
Comment (DanielXiao) [reply packages/tkg-clusterclass-vsphere/bundle/config/upstream/base.yaml:537]: I think we should split this large PR into several small ones. At least those addon code changes should be in a different one. Could you create jira stories or bugs for your code changes so we can break them down and assign them?

File: pkg/v1/providers/config_default.yaml [hunk 18]
```diff
@@ -174,9 +174,9 @@ NSXT_PASSWORD: ""
 #! NSX-T host
 NSXT_MANAGER_HOST: ""
 #! set this to true if NSX-T uses self-signed cert
-NSXT_ALLOW_UNVERIFIED_SSL: "false"
+NSXT_ALLOW_UNVERIFIED_SSL: false
```
Comment (lubronzhan) [inline pkg/v1/providers/config_default.yaml:177]: Why the type was changed? For customer that upgrade to newer version, their config would still be string type, wouldn't it breaks their upgrade?

File: pkg/v1/providers/config_default.yaml [hunk 19]
```diff
@@ -174,9 +174,9 @@ NSXT_PASSWORD: ""
...
 #! set this to true if NSX-T uses remote authentication (authentication done through the vIDM)
-NSXT_REMOTE_AUTH: "false"
+NSXT_REMOTE_AUTH: false
```
Comment (lubronzhan) [inline pkg/v1/providers/config_default.yaml:179]: Same as above

File: addons/controllers/cpi/vspherecpiconfig_utils.go (hunk 2 above)
Comment (lubronzhan) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:88]: >Going forward we expect these to be provided as cluster variables
If that's the case @DanielXiao could you revert this part of the change? Thanks
